src/rdb_protocol/terms/string.cc

```diff
@@ -125,6 +127,55 @@ class match_term_t : public op_term_t {
     virtual const char *name() const { return "match"; }
 };
 
+class format_term_t : public op_term_t {
+public:
+    format_term_t(compile_env_t *env, const raw_term_t &term)
+            : op_term_t(env, term, argspec_t(2)) {}
+private:
+    virtual scoped_ptr_t <val_t> eval_impl(scope_env_t *env, args_t *args, eval_flags_t) const {
+        const std::string templ = args->arg(env, 0)->as_str().to_std();
+        const datum_t datum = args->arg(env, 1)->as_datum();
+
+        std::map <datum_string_t, datum_t> params{};
+        for (size_t i = 0; i < datum.obj_size(); ++i) {
+            const auto datum_pair{datum.get_pair(i)};
+            rcheck_target(&datum_pair.second, datum_pair.second.get_type() == datum_t::R_STR,
+                          base_exc_t::LOGIC,
+                          strprintf("Cannot call `%s` on objects of type `%s`.",
+                                    name(), datum_pair.second.get_type_name().c_str()));
+
+            params.insert(datum_pair);
+        }
+
+        std::string formatted{};
+        for (size_t pos = 0; pos < templ.size(); ++pos) {
+            if (templ[pos] == '{') {
+                size_t param_end {1};
+                std::string param {};
+
+                while (pos + param_end < templ.size() && templ[pos + param_end] != '}') {
+                    param.push_back(templ[pos + param_end]);
+                    ++param_end;
+                }
```

Comment:
A bug here is that if we reach the end of the format string without reaching a closing brace, we'd still try doing the formatting.

Also, if we have a string with extra curly braces, like `{abc{def}` it would try formatting with `abc{def`.  This is probably a user error.  You might want to carefully specify what characters are allowed inside a formatting specifier.

Reply:
Similarly, we probably want to prohibit stray `}` characters in the format string.  Or at least think about it.

Reply:
Thinking about it again, allowing missing closing brackets isn't a good idea. Will refactor this too.